packages/editor/src/store/reducer.js

```diff
@@ -211,6 +213,24 @@ const withInnerBlocksRemoveCascade = ( reducer ) => ( state, action ) => {
 	return reducer( state, action );
 };
 
+/**
+ * Higher-order reducer targeting the combined reducer, augmenting the
+ * RESET_BLOCKS action with a list of client IDs that should remain in editor
+ * state as they are being referenced by a reusable block.
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+const withReusableBlockClientIdsOnReset = ( reducer ) => ( state, action ) => {
+	if ( state && action.type === 'RESET_BLOCKS' ) {
+		const reusableBlockClientIds = map( state.reusableBlocks.data, 'clientId' );
+		action = { ...action, reusableBlockClientIds };
```

Comment:
Close, but my expectation is that our reducer doesn't need to change _at all_, and that our modifications to the action here are to bundle in the reusable blocks we want to preserve into the `action.blocks` property. Is that possible?

This ties into sadness evoked at https://github.com/WordPress/gutenberg/pull/11746#pullrequestreview-173921136 in that ideally at some point we can just drop this higher-order reducer without having to surgically remove it from the reducer.

I wish it could remain an enhancer to the `editor` reducer specifically, since we have a pattern for those established already. I guess it's not simple enough to detect which of the blocks are reusable from within editor state? Possible, yes, but only with a (bulky) operation of gathering all `clientIds` from `order` and doing a difference on those found within `byClientId`. Since `RESET_BLOCKS` happens infrequently enough, maybe it's fine to be bulky. I'm not overly concerned about it one way or the other.

Reply:
> Close, but my expectation is that our reducer doesn't need to change at all, and that our modifications to the action here are to bundle in the reusable blocks we want to preserve into the action.blocks property. Is that possible?

That won't work because if we append the reusable blocks that we want to preserve onto `action.blocks` then the reusable blocks will be added into `blocks.order['']` and incorrectly appear in the post.

We need some way of indicating to the `blocks.byClientId` reducer which blocks should be preserved in state while also indicating to the `blocks.order` reducer which blocks should not end up in `blocks.order['']`. This is the role of `action.reusableBlockClientIds` here.

> I wish it could remain an enhancer to the editor reducer specifically, since we have a pattern for those established already. I guess it's not simple enough to detect which of the blocks are reusable from within editor state? Possible, yes, but only with a (bulky) operation of gathering all clientIds from order and doing a difference on those found within byClientId.

A desire to only enhance the `editor` reducer was what led me to my original approach in 0c92c67ac97d151be3f820ccb3d15b72dfde7183 which is similar to what you're describing here. I'd be happy to go back to that approach. One nice thing about it is that it means that the `blocks` reducer doesn't need to "know" what a reusable block is.

Reply:
>I want to stress that regardless of the approach we take here we can revert this entire PR when something like #7453 does eventually land.
>
>Don't be sad, Andrew! 🙂

The cynic in me assumes that a future major refactor is never to come 😅 

I wonder if a compromise might be to leave the `RESET_BLOCKS` case in the switch, even if technically redundant, to allow for this to actually be seamlessly dropped.

Reply:
> I wonder if a compromise might be to leave the `RESET_BLOCKS` case in the switch, even if technically dead code, to allow for this to actually be seamlessly dropped.

Reflecting on this, I'd rather we not have the dead code, and it should become apparent in the expression of the higher-order reducer that the handling of `RESET_BLOCKS` is to be deferred.

It'll require a bit more care and attention on our part if a refactor should come to pass.